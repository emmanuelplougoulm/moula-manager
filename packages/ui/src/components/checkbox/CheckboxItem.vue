<template>
  <div>
    <input
      id="checkbox"
      v-model="checked"
      class="input"
      type="checkbox"
      data-testid="input"
      @change="onChange"
    />
    <label class="label text-big-regular container" :class="labelClass" for="checkbox">
      <Check class="icon" :class="iconClass" />
      <span class="text">{{ text }}</span>
      <ErrorItem class="error" :text="errorText" />
    </label>
  </div>
</template>

<script setup lang="ts">
import ErrorItem from '../error/ErrorItem.vue';
import { ref, computed } from 'vue';
import Check from '../icons/Check.vue';

const props = defineProps({
  text: {
    type: String,
    required: true
  },
  required: {
    type: Boolean,
    required: true
  },
  error: {
    type: String,
    default: ''
  }
});

const checked = ref(false);
const errorText = ref(props.error);

const iconClass = computed(() => ({
  ['icon--checked']: checked.value,
  ['icon--highlighted']: props.required
}));
const labelClass = computed(() => ({
  ['label--highlighted']: !checked.value && props.required
}));

const emit = defineEmits(['update:checked']);
const onChange = () => {
  errorText.value = !checked.value && props.required ? 'Required' : '';
  emit('update:checked', checked.value);
};
</script>

<style scoped>
.input {
  display: none;
}
.label {
  display: grid;
  align-items: center;
  justify-content: flex-start;
  grid-template-columns: 1.5rem 1fr;
  grid-template-areas:
    'icon text'
    'none error';
  color: var(--neutral-20);
  cursor: pointer;
  gap: 0.5rem;
}

.label--highlighted {
  color: var(--primary-30);
}

.icon {
  grid-area: icon;
  width: 1.25rem;
  height: 1.25rem;
  border-radius: 0.25rem;
  transition: all 0.2s;
  border: 0.125rem solid var(--neutral-40);
  background-color: var(--neutral-100);
  color: var(--neutral-100);
}

.icon:hover {
  border: 0.125rem solid var(--primary-30);
}

.icon--checked {
  grid-area: icon;
  border: 0.125rem solid var(--primary-30);
  background-color: var(--primary-30);
}

.icon--highlighted {
  border: 0.125rem solid var(--primary-30);
}

.text {
  grid-area: text;
}

.error {
  grid-area: error;
}
</style>
