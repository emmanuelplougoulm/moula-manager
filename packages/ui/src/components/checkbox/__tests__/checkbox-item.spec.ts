import { mount } from '@vue/test-utils';
import CheckboxItem from '../CheckboxItem.vue';

const inputTestId = "[data-testid='input']";

describe('CheckboxItem tests', () => {
  it('should display the component with the default classes', () => {
    const wrapper = mount(CheckboxItem, {
      props: {
        text: 'Text',
        required: false
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should display the component with the highlighted classes', () => {
    const wrapper = mount(CheckboxItem, {
      props: {
        text: 'Text',
        required: true
      }
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should emit an update:checked event if the box is checked', async () => {
    const wrapper = mount(CheckboxItem, {
      props: {
        text: 'Text',
        required: false
      }
    });
    const input = wrapper.find(inputTestId);
    await input.setValue(true);

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.emitted()).toHaveProperty('update:checked');
  });

  it('should not display an error message if the checkbox is checked', async () => {
    const wrapper = mount(CheckboxItem, {
      props: {
        text: 'Text',
        required: true,
        error: 'This field is required'
      }
    });
    const input = wrapper.find(inputTestId);
    await input.setValue(true);

    expect(wrapper.html()).toMatchSnapshot();
  });
});

it('should display an error message if the checkbox is required and unchecked', async () => {
  const wrapper = mount(CheckboxItem, {
    props: {
      text: 'Text',
      required: true,
      error: 'This field is required'
    }
  });
  const input = wrapper.find(inputTestId);
  await input.setValue(true);
  await input.setValue(false);

  expect(wrapper.html()).toMatchSnapshot();
});
