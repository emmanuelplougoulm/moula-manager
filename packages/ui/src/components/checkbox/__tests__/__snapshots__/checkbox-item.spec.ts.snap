// Vitest Snapshot v1

exports[`CheckboxItem tests > should display the component with the default classes 1`] = `
"<div data-v-58fd6536=\\"\\"><input data-v-58fd6536=\\"\\" id=\\"checkbox\\" class=\\"input\\" type=\\"checkbox\\" data-testid=\\"input\\"><label data-v-58fd6536=\\"\\" class=\\"label text-big-regular container\\" for=\\"checkbox\\">
    <div data-v-b9dd0015=\\"\\" data-v-58fd6536=\\"\\" class=\\"container icon\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
        <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
      </svg></div><span data-v-58fd6536=\\"\\" class=\\"text\\">Text</span>
    <!--v-if-->
  </label></div>"
`;

exports[`CheckboxItem tests > should display the component with the highlighted classes 1`] = `
"<div data-v-58fd6536=\\"\\"><input data-v-58fd6536=\\"\\" id=\\"checkbox\\" class=\\"input\\" type=\\"checkbox\\" data-testid=\\"input\\"><label data-v-58fd6536=\\"\\" class=\\"label text-big-regular container label--highlighted\\" for=\\"checkbox\\">
    <div data-v-b9dd0015=\\"\\" data-v-58fd6536=\\"\\" class=\\"container icon icon--highlighted\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
        <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
      </svg></div><span data-v-58fd6536=\\"\\" class=\\"text\\">Text</span>
    <!--v-if-->
  </label></div>"
`;

exports[`CheckboxItem tests > should emit an update:checked event if the box is checked 1`] = `
"<div data-v-58fd6536=\\"\\"><input data-v-58fd6536=\\"\\" id=\\"checkbox\\" class=\\"input\\" type=\\"checkbox\\" data-testid=\\"input\\"><label data-v-58fd6536=\\"\\" class=\\"label text-big-regular container\\" for=\\"checkbox\\">
    <div data-v-b9dd0015=\\"\\" data-v-58fd6536=\\"\\" class=\\"container icon icon--checked\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
        <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
      </svg></div><span data-v-58fd6536=\\"\\" class=\\"text\\">Text</span>
    <!--v-if-->
  </label></div>"
`;

exports[`CheckboxItem tests > should not display an error message if the checkbox is checked 1`] = `
"<div data-v-58fd6536=\\"\\"><input data-v-58fd6536=\\"\\" id=\\"checkbox\\" class=\\"input\\" type=\\"checkbox\\" data-testid=\\"input\\"><label data-v-58fd6536=\\"\\" class=\\"label text-big-regular container\\" for=\\"checkbox\\">
    <div data-v-b9dd0015=\\"\\" data-v-58fd6536=\\"\\" class=\\"container icon icon--checked icon--highlighted\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
        <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
      </svg></div><span data-v-58fd6536=\\"\\" class=\\"text\\">Text</span>
    <!--v-if-->
  </label></div>"
`;

exports[`should display an error message if the checkbox is required and unchecked 1`] = `
"<div data-v-58fd6536=\\"\\"><input data-v-58fd6536=\\"\\" id=\\"checkbox\\" class=\\"input\\" type=\\"checkbox\\" data-testid=\\"input\\"><label data-v-58fd6536=\\"\\" class=\\"label text-big-regular container label--highlighted\\" for=\\"checkbox\\">
    <div data-v-b9dd0015=\\"\\" data-v-58fd6536=\\"\\" class=\\"container icon icon--highlighted\\"><svg data-v-b9dd0015-s=\\"\\" xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"48\\" height=\\"48\\" viewBox=\\"0 96 960 960\\">
        <path d=\\"M378 823.566 140.434 586l57.131-57.131L378 709.304l383.435-383.435L818.566 383 378 823.566Z\\"></path>
      </svg></div><span data-v-58fd6536=\\"\\" class=\\"text\\">Text</span><span data-v-c1d8433f=\\"\\" data-v-58fd6536=\\"\\" id=\\"error-item\\" data-testid=\\"error-item\\" class=\\"text-small-regular error\\">Required</span>
  </label></div>"
`;
